0000-notes-on-oer-encoding/0000-notes-on-oer-encoding.md

```diff
@@ -0,0 +1,190 @@
+
+We chose CANONICAL-OER instead of BASIC-OER because we wanted a canonical encoding format so that Interledger encodings could be used for cryptographic operations like hashing and signing.
+
+## Basics
+
+### Octet strings
+
+All Interledger formats are byte-aligned messages and should be passed as octet strings, i.e. sequences of bytes. The correct type for this would be a `Buffer` in JavaScript, `byte[]` in Java, `bytes` in Python, etc.
+
+### Extensibility
+
+For most Interledger formats, we require that implementations MUST ignore extra bytes at the end of the message. This behavior allows additional fields to be added in the future that are automatically ignored by existing implementations.
+
+## Types
+
+### Fixed-length unsigned integers
+
+All fixed-length unsigned integers are encoded in big-endian, least-significant bit last. Interledger encoding provides the following fixed-length unsigned integers:
+
+| Type | Size | Range |
+| :--- | :--- | :--- |
+| UInt8 | 1 byte | 0 .. 255 (2^8 - 1) |
+| UInt16 | 2 bytes | 0 .. 65535 (2^16 - 1) |
+| UInt32 | 4 bytes | 0 .. 4294967295 (2^32 - 1) |
+| UInt64 | 8 bytes | 0 .. 18446744073709551615 (2^64 - 1) |
+| UInt128 | 16 bytes | 0 .. 2^128 - 1 |
+| UInt160 | 20 bytes | 0 .. 2^160 - 1 |
+| UInt192 | 24 bytes | 0 .. 2^192 - 1 |
+| UInt224 | 28 bytes | 0 .. 2^224 - 1 |
+| UInt256 | 32 bytes | 0 .. 2^256 - 1 |
+| UInt384 | 48 bytes | 0 .. 2^384 - 1 |
+| UInt512 | 64 bytes | 0 .. 2^512 - 1 |
+
+### Fixed-length signed integers
+
+All fixed-length signed integers are encoded in big-endian, least-significant bit last. Negative values are encoded using 2's complement binary encoding.
+
+| Type | Size | Range |
+| :--- | :--- | :--- |
+| Int8 | 1 byte | -128 .. 127 |
+| Int16 | 2 bytes | -32768 .. 32767 |
+| Int32 | 4 bytes | -2147483648 .. 2147483647 |
+| Int64 | 8 bytes | -9223372036854775808 .. 9223372036854775807 |
+
+### Fixed-length floating point numbers
+
+Floating point values in Interledger are encoded using IEEE 754 binary floating point.
+
+| Type | Size | IEEE 754 name |
+| :--- | :--- | :--- |
+| Float32 | 4 bytes | binary32 |
+| Float64 | 8 bytes | binary64 |
+
+### Fixed-length octet string
+
+A fixed length octet string is encoded as itself with no added prefix or suffix.
+
+### Length determinant
+
+All variable-length fields are prefixed with a length determinant. The length determinant can be either short form or long form.
+
+#### Short form
+
+A single byte in the range 0 .. 127, denoting the length.
+
+#### Long form
+
+A single byte in the range 128 .. 255, denoting the value *128 + n*, where *n* is the length-of-length. This is followed by an unsigned integer of *n* bytes denoting the length. The length MUST NOT contain leading zeros and MUST be greater than 127.
```

Comment:
Might be a little confusing here; the length-of-length needn't be greater than 127, but the length-of-length long integer containing the length of the data must be greater than 127